bundles/automation/org.eclipse.smarthome.automation.module.script.rulesupport/src/main/java/org/eclipse/smarthome/automation/module/script/rulesupport/internal/ScriptedCustomModuleHandlerFactory.java

```diff
@@ -0,0 +1,87 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.automation.module.script.rulesupport.internal;
+
+import java.util.Collection;
+import java.util.Dictionary;
+import java.util.HashMap;
+import java.util.Hashtable;
+
+import org.eclipse.smarthome.automation.Module;
+import org.eclipse.smarthome.automation.handler.ModuleHandler;
+import org.eclipse.smarthome.automation.handler.ModuleHandlerFactory;
+import org.eclipse.smarthome.automation.module.script.rulesupport.shared.ScriptedHandler;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ScriptedCustomModuleHandlerFactory} is used in combination with the
+ * {@link ScriptedCustomModuleTypeProvider} to allow scripts to define custom types in the RuleEngine. These
+ * registered types can then be used publicly from any Rule-Editor.
+ *
+ * This class provides the handlers from the script to the RuleEngine. As Jsr223 languages have different needs, it
+ * allows these handlers to be defined in different ways.
+ *
+ * @author Simon Merschjohann
+ *
+ */
+public class ScriptedCustomModuleHandlerFactory extends AbstractScriptedModuleHandlerFactory {
+    private Logger logger = LoggerFactory.getLogger(ScriptedCustomModuleHandlerFactory.class);
+
+    private final HashMap<String, ScriptedHandler> typesHandlers = new HashMap<>();
+
+    private ServiceRegistration<?> bmhfReg;
+
+    private static ScriptedCustomModuleHandlerFactory instance;
+
+    public static ScriptedCustomModuleHandlerFactory get() {
+        return instance;
+    }
+
+    @Override
+    public void activate(BundleContext bundleContext) {
+        super.activate(bundleContext);
+
+        bmhfReg = bundleContext.registerService(ModuleHandlerFactory.class.getName(), this, null);
+
+        instance = this;
+    }
+
+    @Override
+    public Collection<String> getTypes() {
+        return typesHandlers.keySet();
+    }
+
+    @Override
+    protected ModuleHandler internalCreate(Module module, String ruleUID) {
+        ScriptedHandler scriptedHandler = typesHandlers.get(module.getTypeUID());
+
+        return getModuleHandler(module, scriptedHandler);
+    }
+
+    public void addModuleHandler(String uid, ScriptedHandler scriptedHandler) {
+        typesHandlers.put(uid, scriptedHandler);
+
+        updateModuleHandlerOffer();
+    }
+
+    public void removeModuleHandler(String uid) {
+        typesHandlers.remove(uid);
+
+        updateModuleHandlerOffer();
+    }
+
+    private void updateModuleHandlerOffer() {
+        Dictionary<String, Object> properties = new Hashtable<String, Object>();
+
+        properties.put("handlers", typesHandlers.values());
```

Comment:
I just digged a bit further.

> This is done to trigger "updatedModuleHandlerFactory"

I noticed that the PR that introduced this method actually increased the DS declaration from 1.1.0 to 1.2.0, which is part of the OSGi 4.3 spec - unfortunately, we have to be compatible with 4.2 as well, so we have to stick to DS 1.1.0, so I think we will have to revert that change - @danchom Is this correct?

Without this method, setting the properties might also not be the best approach; instead of modifying, we should probable unregister and re-register it.

For this, I would agree that it is simplest to do the registration in activate() and keep the `ServiceRegistration` instance, which can be used here in `updateModuleHandlerOffer` to unregister the service. Re-registration would then give us a new `ServiceRegistration` instance, which can be stored in the field.

Does that make sense?